paasta_tools/long_running_service_tools.py

```diff
@@ -72,6 +74,7 @@ class LongRunningServiceConfigDict(InstanceConfigDict, total=False):
     registrations: List[str]
     replication_threshold: int
     bounce_start_deadline: float
+    smartstack: dict
```

Comment:
Where does this data come from? `LongRunningServiceConfigDict` (and its subclasses like `KubernetesDeploymentConfigDict`) represent the contents of one instance definition in a `kubernetes-clustername.yaml`.

I'm assuming this `smartstack` key is meant to represent an entry in smartstack.yaml? If so, you probably want to look at `load_service_namespace_config`, which gives you a `ServiceNamespaceConfig`, which represents an entry in smartstack.yaml.

Reply:
@cuza found that every `LongRunningServiceConfig` object being pulled into `paasta validate` already had a `config_dict`  (A `LongRunningServiceConfigDict`) that contained a `smartstack` attribute.  Upon reviewing the code it looks like every subclass of LongRunningServiceConfig sets `config_dict` by using [service_configuration_lib.read_service_configuration](https://github.com/Yelp/service_configuration_lib/blob/master/service_configuration_lib/__init__.py#L158), which pulls from multiple yelpsoa-config files, including smartstack.yaml, and so *will* have this information, albeit not as strictly in our `ServiceNamespaceConfig` format, but with enough information to _validate_ that registrations are contained within smartstack.yaml.

Since the validator handles the case where we do not have the `smartstack` key, this seems like it should work.

Reply:
Ah, interesting. I think it might be better to do this by defining another TypedDict subclass in service_configuration_lib that represents what `service_configuration_lib.read_service_configuration_from_dir()` can return; then make `InstanceConfigDict` (or `LongRunningServiceConfigDict` or `KubernetesDeploymentConfigDict`) inherit from that. This way, it's explicit that the magic `smartstack` key (or `port`/`monitoring`/`deploy`/`data`/`dependencies`) come from service_configuration_lib, not from being loaded from `kubernetes-whatever.yaml`.

You could put the new type in service_configuration_lib (better) or define it in the paasta repo (easier).